### PR TITLE
Extend WO engine functional tests

### PR DESCRIPTION
**Pull request by arutk**
1. Adjust test to work with multiple cache lines
2. Extend test workset to cover larger I/O
3. Run the test for reads in cache modes other than WO as well.

Signed-off-by: Adam Rutkowski <adam.j.rutkowski@intel.com>

### CODE REVIEW

File: tests/functional/tests/engine/test_wo.py [hunk 1]
```diff
@@ -165,49 +305,88 @@ def test_wo_read_data_consistency(pyocf_ctx):
...
-                assert 0 == io_to_exp_obj(
-                    core, WORKSET_OFFSET + START, size, result_b, START, IoDir.READ
-                ), "error reading in WO mode: S={}, start={}, end={}, insert_order={}".format(
-                    S, s, e, insert_order
+        # get up to 32 randomly selected pairs of (start,end) sectors
+        # 32 is enough to cover all combinations for 4K and 8K cacheline size
+        io_ranges = []
+        for r in product(start_sec, end_sec):
+            start = r[0]
+            end = r[1]
```
Comment (robertbaldyga) [inline tests/functional/tests/engine/test_wo.py]:
```python
for start, end in product(start_sec, end_sec):
```

File: tests/functional/tests/engine/test_wo.py [hunk 2]
```diff
@@ -165,49 +305,88 @@ def test_wo_read_data_consistency(pyocf_ctx):
...
+            if start > end:
+                continue
+            io_ranges += [r]
+        random.shuffle(io_ranges)
+        io_ranges = io_ranges[:32]
```
Comment (robertbaldyga) [inline tests/functional/tests/engine/test_wo.py]: Same here. Random values break reproducibility. Please parametrize test with random seed.

File: tests/functional/tests/engine/test_wo.py [hunk 3]
```diff
@@ -55,106 +65,236 @@ def sector_to_region(sector, region_start):
...
+        + [I]
+        + [D for i in range(CLS // 2)]
+        + [D, I, I],
+    ]
+
+    fixed_statuses_count = len(region_statuses)
+
+    # add randomly generated sector statuses
+    for i in range(ITRATION_COUNT - fixed_statuses_count):
+        region_statuses += [[SectorStatus(random.randrange(3)) for i in range(len(region_start))]]
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply tests/functional/tests/engine/test_wo.py]: or `[[random.choice(list(SectorStatus)) (...)`

File: tests/functional/tests/engine/test_read.py [hunk 4]
```diff
@@ -384,9 +390,10 @@ def test_wo_read_data_consistency(pyocf_ctx, cls):
...
-            # WO is not supposed to clean dirty data
-            assert (
-                core_device.get_stats()[IoDir.WRITE] == 0
-            ), "unexpected write to core device, region_state={}, start={}, end={}, insert_order = {}\n".format(
-                region_state, start, end, insert_order
-            )
+            if cache_mode == CacheMode.WO:
+                # WO is not supposed to clean dirty data
+                assert (
+                    core_device.get_stats()[IoDir.WRITE] == 0
```
[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/tests/engine/test_read.py:410]: true, but it's a lot of work to put it somewhere else...

[the remark this answers is not in the document]
Comment (arutk) [reply tests/functional/tests/engine/test_read.py:408]: I suggest to leave this as is

**Review by robertbaldyga — DISMISSED** [review]

**Review by robertbaldyga — APPROVED** [review]

**Review by robertbaldyga — APPROVED** [review]

**Comment by arutk** [comment]
retest this please

**Review by mmichal10 — APPROVED** [review]

**Pull request merged**